src/net.rs

```diff
@@ -124,14 +124,35 @@ pub fn send_error_package<W:Write>(stream: &mut W, err: ErrorCode) -> io::Result
 /// commands that are not sent as query.
 #[derive(RustcEncodable, RustcDecodable)]
 #[repr(u8)]
+#[derive(PartialEq, Eq, Debug)]
 pub enum Command {
     Quit,
     Ping,
     Query(String),
+    Unknown,
     // Shutdown,
     // Statistics,
 }
 
+/// read sent bytes, extract the kind of command
+pub fn read_commands<R:Read + Write>(stream: &mut R) -> Result<Command, io::Error> {
+    
+    // read the first byte for code numeric value
+    let status = try!(stream.read_u8());
+    if status != Cnv::CommandPkg as u8 {
+        //send error_package
+        try!(send_error_package(stream, ErrorCode::UnexpecPkgErr));
+        return Ok(Command::Unknown)
+    }
+    
+    // second  4 bytes is the kind of command
+    let command_decode = decode_from::<R,Command>(stream, SizeLimit::Bounded(4096));
```

Comment:
If I'm not mistaken, you can remove the generic parameters there (`::<R,Command>`)..